src/algebra/module/linear_map.lean

```diff
@@ -0,0 +1,419 @@
+  M₂ →ₗ[R] M :=
+by dsimp [left_inverse, function.right_inverse] at h₁ h₂; exact
+  ⟨g, λ x y, by { rw [← h₁ (g (x + y)), ← h₁ (g x + g y)]; simp [h₂] },
+      λ a b, by { rw [← h₁ (g (a • b)), ← h₁ (a • g b)]; simp [h₂] }⟩
+
+end add_comm_monoid
+
+section add_comm_group
+
+variables [semiring R] [add_comm_group M] [add_comm_group M₂]
+variables {semimodule_M : semimodule R M} {semimodule_M₂ : semimodule R M₂}
+variables (f : M →ₗ[R] M₂)
+
+@[simp] lemma map_neg (x : M) : f (- x) = - f x :=
+f.to_add_monoid_hom.map_neg x
+
+@[simp] lemma map_sub (x y : M) : f (x - y) = f x - f y :=
+f.to_add_monoid_hom.map_sub x y
+
+instance : is_add_group_hom f :=
+{ map_add := map_add f}
+
+end add_comm_group
+
+end linear_map
+
+namespace is_linear_map
+
+section add_comm_monoid
+variables [semiring R] [add_comm_monoid M] [add_comm_monoid M₂]
+variables [semimodule R M] [semimodule R M₂]
+include R
+
+/-- Convert an `is_linear_map` predicate to a `linear_map` -/
+def mk' (f : M → M₂) (H : is_linear_map R f) : M →ₗ M₂ := ⟨f, H.1, H.2⟩
+
+@[simp] theorem mk'_apply {f : M → M₂} (H : is_linear_map R f) (x : M) :
+  mk' f H x = f x := rfl
+
+lemma is_linear_map_smul {R M : Type*} [comm_semiring R] [add_comm_monoid M] [semimodule R M]
+  (c : R) :
+  is_linear_map R (λ (z : M), c • z) :=
+begin
+  refine is_linear_map.mk (smul_add c) _,
+  intros _ _,
+  simp only [smul_smul, mul_comm]
+end
+
+--TODO: move
+lemma is_linear_map_smul' {R M : Type*} [semiring R] [add_comm_monoid M] [semimodule R M] (a : M) :
+  is_linear_map R (λ (c : R), c • a) :=
+is_linear_map.mk (λ x y, add_smul x y a) (λ x y, mul_smul x y a)
+
+variables {f : M → M₂} (lin : is_linear_map R f)
+include M M₂ lin
+
+lemma map_zero : f (0 : M) = (0 : M₂) := (lin.mk' f).map_zero
+
+end add_comm_monoid
+
+section add_comm_group
+
+variables [semiring R] [add_comm_group M] [add_comm_group M₂]
+variables [semimodule R M] [semimodule R M₂]
+include R
+
+lemma is_linear_map_neg :
+  is_linear_map R (λ (z : M), -z) :=
+is_linear_map.mk neg_add (λ x y, (smul_neg x y).symm)
+
+variables {f : M → M₂} (lin : is_linear_map R f)
+include M M₂ lin
+
+lemma map_neg (x : M) : f (- x) = - f x := (lin.mk' f).map_neg x
+
+lemma map_sub (x y) : f (x - y) = f x - f y := (lin.mk' f).map_sub x y
+
+end add_comm_group
+
+end is_linear_map
+
+/-- Ring of linear endomorphismsms of a module. -/
+abbreviation module.End (R : Type u) (M : Type v)
+  [semiring R] [add_comm_monoid M] [semimodule R M] := M →ₗ[R] M
+
+/-- Reinterpret an additive homomorphism as a `ℤ`-linear map. -/
+def add_monoid_hom.to_int_linear_map [add_comm_group M] [add_comm_group M₂] (f : M →+ M₂) :
+  M →ₗ[ℤ] M₂ :=
+⟨f, f.map_add, f.map_int_module_smul⟩
+
+/-- Reinterpret an additive homomorphism as a `ℚ`-linear map. -/
+def add_monoid_hom.to_rat_linear_map [add_comm_group M] [vector_space ℚ M]
+  [add_comm_group M₂] [vector_space ℚ M₂] (f : M →+ M₂) :
+  M →ₗ[ℚ] M₂ :=
+⟨f, f.map_add, f.map_rat_module_smul⟩
```

Comment:
Would this be better or worse as:
```suggestion
{map_smul ' := f.map_rat_module_smul, ..f}
```

Reply:
I guess moving definitions is a good excuse to clean them up a bit. It does look better, thanks!